src/containers/Interfaces/AlterForm.js

```diff
@@ -0,0 +1,217 @@
+import React, { Component } from 'react';
+import cx from 'classnames';
+import PropTypes from 'prop-types';
+import { bindActionCreators, compose } from 'redux';
+import { connect } from 'react-redux';
+import { isValid, isSubmitting, submit } from 'redux-form';
+import Swiper from 'react-id-swiper';
+import { Scroller, ProgressBar } from '../../components';
+import Node from '../Node';
+import { actionCreators as sessionsActions } from '../../ducks/modules/sessions';
+import { nodeAttributesProperty, nodePrimaryKeyProperty } from '../../ducks/modules/network';
+import { makeNetworkNodesForType } from '../../selectors/interface';
+import { protocolForms } from '../../selectors/protocol';
+
+import { Form } from '../';
+import { getItemComponent } from './Information';
+import { getCSSVariableAsNumber } from '../../ui/utils/CSSVariables';
+
+class AlterForm extends Component {
+  constructor(props) {
+    super(props);
+    this.swipeRef = React.createRef();
+    this.state = {
+      activeIndex: 0,
+    };
+  }
+
+  onSlideTransition = () => {
+    if (this.swipeRef.current && this.swipeRef.current.swiper) {
+      const submitIndex = this.swipeRef.current.swiper.previousIndex;
+      // try to submit (on fail, form is "touched" and errors can show)
+      this.props.submitForm(this.getNodeFormName(submitIndex));
+      if (!this.formSubmitAllowed(submitIndex)) {
+        this.swipeRef.current.swiper.slideTo(submitIndex,
+          getCSSVariableAsNumber('--animation-duration-fast-ms'), false);
+      } else {
+        this.setState({
+          activeIndex: this.swipeRef.current.swiper.activeIndex,
+        });
+      }
+    }
+  }
+
+  getNodeFormName = activeIndex => `NODE_FORM_${activeIndex}`;
+
+  formSubmitAllowed = index => (
+    this.swipeRef && this.swipeRef.current.swiper &&
+    this.props.formEnabled(this.getNodeFormName(index))
+  );
+
+  isStageBeginning = () => (
+    this.swipeRef && this.formSubmitAllowed(this.swipeRef.current.swiper.activeIndex) &&
+    this.swipeRef.current.swiper.activeIndex === 0
+  );
+
+  isStageEnding = () => (
+    this.swipeRef && this.formSubmitAllowed(this.swipeRef.current.swiper.activeIndex) &&
+    this.swipeRef.current.swiper.activeIndex === this.props.stageNodes.length
+  );
+
+  clickNext = () => {
+    if (this.swipeRef && this.swipeRef.current) {
+      if (this.swipeRef.current.swiper.isEnd) {
```

Comment:
Given the `!this.getInterfaceRef().isStageEnding()`, is it possible to reach this statement? The functionality seems to be handled by `clickPrevious`.

It would be nice to have this logic encapsulated in a single callback rather than having both 'next' and 'last' handlers here, but that might be out of scope for this PR.

Reply:
It was possible...if you tried to advance but it wasn't a valid form and so wasn't ready to move to the next Stage. I combined the logic and simplified a bit because we moved away from supporting swiping.